packages/std/src/query.rs

```diff
@@ -28,6 +28,29 @@ pub enum BankQuery {
     AllBalances { address: HumanAddr },
 }
 
+/// A trait that is required to avoid conflicts with other query types like BankQuery and WasmQuery
+/// in generic implementations.
+/// You need to implement it in your custom query type.
+///
+/// # Examples
+///
+/// ```
+/// # use cosmwasm_std::CustomQuery;
+/// # use schemars::JsonSchema;
+/// # use serde::{Deserialize, Serialize};
+/// #[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+/// #[serde(rename_all = "snake_case")]
+/// pub enum MyCustomQuery {
+///     Ping {},
+///     Capitalized { text: String },
+/// }
+///
+/// impl CustomQuery for MyCustomQuery {}
+/// ```
+pub trait CustomQuery: Serialize {}
```

Comment:
Why isn't this `Serialize + DeserializedOwned` as that is what you use above a few places?

Reply:
And you replace `T: Clone + fmt::Debug + PartialEq + JsonSchema` with `C: CustomQuery`.

I am a bit confused how this all works (it has also been a long time since I wrote this code)

Reply:
I kept it as minimal as possible and add trait restructions carefilly and it turned out that `Clone + fmt::Debug + PartialEq + JsonSchema` is not required. Maybe  Rust infers this for QueryRequest somehow. You can do

```rust
#[derive(Serialize)]
struct Foo {}

impl CustomQuery for Foo {}

pub type FooQuery = QueryRequest<Foo>;
```

`DeserializeOwned` is something needed in the mock querier only. In the rest of the code, the query is only serialized in Rust land.

Reply:
`Clone + fmt::Debug + PartialEq + JsonSchema` is needed when the custom query is part of a public facing message, like in

```rust
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
#[serde(rename_all = "snake_case")]
pub enum QueryMsg {
    // …
    Chain {
        request: QueryRequest<SpecialQuery>,
    },
}
```

but when the contract uses it internally, it is not needed.

Reply:
Nice finds, thank you